File: app/src/main/java/es/etologic/mahjongscoring2/domain/use_cases/UpdateRoundUseCase.java
package es.etologic.mahjongscoring2.domain.use_cases;

import es.etologic.mahjongscoring2.data.repository.DataProvider;
import es.etologic.mahjongscoring2.domain.DataSource;
import es.etologic.mahjongscoring2.domain.entities.Game;
import es.etologic.mahjongscoring2.domain.entities.Round;
import es.etologic.mahjongscoring2.domain.threading.UseCase;

public class UpdateRoundUseCase extends
        UseCase<UpdateRoundUseCase.RequestValues, UpdateRoundUseCase.ResponseValue> {

    private final DataProvider dataProvider;

    public UpdateRoundUseCase(DataSource dataSource) {
        this.dataProvider = (DataProvider)dataSource;
    }

    @Override
    protected void executeUseCase(RequestValues requestValues) {
        boolean wasUpdated = dataProvider.updateRound(requestValues.getRound());
        if(wasUpdated) getUseCaseCallback().onSuccess(new ResponseValue());
        else getUseCaseCallback().onError(null);
    }

    public static final class RequestValues implements UseCase.RequestValues {

        private Round round;

        public RequestValues(Round round) {
            this.round = round;
        }

        public Round getRound() {
            return round;
        }
    }

    public static final class ResponseValue implements UseCase.ResponseValue {}
}

File: app/src/main/java/es/etologic/mahjongscoring2/domain/use_cases/UpdateRoundUseCase.kt
package es.etologic.mahjongscoring2.domain.use_cases

import es.etologic.mahjongscoring2.data.repositories.GamesRepository
import es.etologic.mahjongscoring2.data.repositories.RoundsRepository
import es.etologic.mahjongscoring2.domain.model.GameWithRounds
import es.etologic.mahjongscoring2.domain.model.Round
import io.reactivex.Single
import javax.inject.Inject

class UpdateRoundUseCase @Inject
constructor(private val roundsRepository: RoundsRepository, private val gamesRepository: GamesRepository) {
    
    internal fun getGameRounds(gameId: Long): Single<List<Round>> {
        return roundsRepository.getRoundsByGame(gameId)
    }
    
    internal fun addRound(round: Round): Single<GameWithRounds> = roundsRepository.insertOne(round)
        .flatMap { gamesRepository.getOneWithRounds(round.gameId.toLong()) } //FixMe: ¿ToLong???
    
    private fun saveRonRound(requestedPoints: Int) {
//        mCurrentRound.handPoints = requestedPoints
//        mCurrentRound.winnerInitialPosition = selectedPlayerSeat.initialSeat
//        val discarderInitialSeat = GameRoundsUtils.getPlayerInitialSeatByCurrentSeat(mDiscarderCurrentSeat, mCurrentRound.roundId)
//        mCurrentRound.discarderInitialPosition = discarderInitialSeat
//        mCurrentRound.setAllPlayersRonPoints(selectedPlayerSeat.initialSeat, requestedPoints, discarderInitialSeat)
//        saveCurrentRoundAndStartNext()
    }
    
    private fun saveTsumoRound(requestedPoints: Int) {
//        mCurrentRound.handPoints = requestedPoints
//        mCurrentRound.winnerInitialPosition = selectedPlayerSeat.initialSeat
//        mCurrentRound.setAllPlayersTsumoPoints(selectedPlayerSeat.initialSeat, requestedPoints)
//        saveCurrentRoundAndStartNext()
    }
    
    private fun saveDrawRound() {
//        mCurrentRound.applyAllPlayersPenalties()
//        //        mCurrentRound.setRoundDuration();
//        disposables.add(
//            updateRoundsUseCase.addRound(mCurrentRound)
//                .subscribeOn(Schedulers.io())
//                .doOnSubscribe { progressState.postValue(SHOW) }
//                .doOnEvent { _, _ -> progressState.postValue(HIDE) }
//                .subscribe({ this.getGameSuccess(it) }, { error.postValue(it) })
//        )
    }
}
